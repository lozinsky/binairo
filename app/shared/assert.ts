class AssertionError extends Error {
  constructor(message?: string) {
    super(message);

    this.name = 'AssertionError';
  }
}

export function assert(condition: boolean, message?: string): asserts condition {
  if (!condition) {
    throw new AssertionError(message ?? 'Expected condition to be true');
  }
}

export function unreachable(message?: string): never {
  throw new AssertionError(message ?? 'Reached unreachable code');
}
