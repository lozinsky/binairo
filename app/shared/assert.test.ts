import { expect, test } from 'vitest';

import { assert, unreachable } from './assert';

test('does not throw if condition is "true"', () => {
  expect(() => {
    assert(true);
  }).not.toThrow();
});

test('throws if condition is "false"', () => {
  expect(() => {
    assert(false);
  }).toThrowErrorMatchingSnapshot();
});

test('throws if condition is "false" with message', () => {
  expect(() => {
    assert(false, 'Message');
  }).toThrowErrorMatchingSnapshot();
});

test('unreachable always throws', () => {
  expect(() => {
    unreachable();
  }).toThrowErrorMatchingSnapshot();
});
