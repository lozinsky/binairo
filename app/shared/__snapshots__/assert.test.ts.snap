// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`throws if condition is "false" 1`] = `[AssertionError: Expected condition to be true]`;

exports[`throws if condition is "false" with message 1`] = `[AssertionError: Message]`;

exports[`unreachable always throws 1`] = `[AssertionError: Reached unreachable code]`;
